homeassistant/components/homematic.py

```diff
@@ -0,0 +1,528 @@
+                              "Motion", "MotionV2", "RemoteMotion"],
+    DISCOVER_ROLLERSHUTTER: ["Blind"]
+}
+
+HM_IGNORE_DISCOVERY_NODE = [
+    "ACTUAL_TEMPERATURE"
+]
+
+HM_ATTRIBUTE_SUPPORT = {
+    "LOWBAT": ["Battery", {0: "High", 1: "Low"}],
+    "ERROR": ["Sabotage", {0: "No", 1: "Yes"}],
+    "RSSI_DEVICE": ["RSSI", {}],
+    "VALVE_STATE": ["Valve", {}],
+    "BATTERY_STATE": ["Battery", {}],
+    "CONTROL_MODE": ["Mode", {0: "Auto", 1: "Manual", 2: "Away", 3: "Boost"}],
+    "POWER": ["Power", {}],
+    "CURRENT": ["Current", {}],
+    "VOLTAGE": ["Voltage", {}]
+}
+
+_HM_DISCOVER_HASS = None
+_LOGGER = logging.getLogger(__name__)
+
+
+# pylint: disable=unused-argument
+def setup(hass, config):
+    """Setup the Homematic component."""
+    global HOMEMATIC, HOMEMATIC_AUTODETECT, _HM_DISCOVER_HASS
+
+    from pyhomematic import HMConnection
+
+    local_ip = config[DOMAIN].get("local_ip", None)
+    local_port = config[DOMAIN].get("local_port", 8943)
+    remote_ip = config[DOMAIN].get("remote_ip", None)
+    remote_port = config[DOMAIN].get("remote_port", 2001)
+    autodetect = config[DOMAIN].get("autodetect", False)
+
+    if remote_ip is None or local_ip is None:
+        _LOGGER.error("Missing remote CCU/Homegear or local address")
+        return False
+
+    # Create server thread
+    HOMEMATIC_AUTODETECT = autodetect
+    _HM_DISCOVER_HASS = hass
+    HOMEMATIC = HMConnection(local=local_ip,
+                             localport=local_port,
+                             remote=remote_ip,
+                             remoteport=remote_port,
+                             systemcallback=system_callback_handler,
+                             interface_id="homeassistant")
+
+    # Start server thread, connect to peer, initialize to receive events
+    HOMEMATIC.start()
+
+    # Stops server when Homeassistant is shutting down
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, HOMEMATIC.stop)
+    hass.config.components.append(DOMAIN)
+
+    return True
+
+
+# pylint: disable=too-many-branches
+def system_callback_handler(src, *args):
+    """Callback handler."""
+    if src == 'newDevices':
+        # pylint: disable=unused-variable
+        (interface_id, dev_descriptions) = args
+        key_dict = {}
+        # Get list of all keys of the devices (ignoring channels)
+        for dev in dev_descriptions:
+            key_dict[dev['ADDRESS'].split(':')[0]] = True
+        # Connect devices already created in HA to pyhomematic and
+        # add remaining devices to list
+        devices_not_created = []
+        for dev in key_dict:
+            try:
+                if dev in HOMEMATIC_DEVICES:
+                    for hm_element in HOMEMATIC_DEVICES[dev]:
+                        hm_element.link_homematic()
+                else:
+                    devices_not_created.append(dev)
+            # pylint: disable=broad-except
+            except Exception as err:
```

Comment:
It's fine to be fixed in a future pull request

On Sat, Jun 25, 2016, 07:06 Daniel <notifications@github.com> wrote:

> In homeassistant/components/homematic.py
> <https://github.com/home-assistant/home-assistant/pull/2368#discussion_r68488956>
> :
>
> > +        key_dict = {}
> > +        # Get list of all keys of the devices (ignoring channels)
> > +        for dev in dev_descriptions:
> > +            key_dict[dev['ADDRESS'].split(':')[0]] = True
> > +        # Connect devices already created in HA to pyhomematic and
> > +        # add remaining devices to list
> > +        devices_not_created = []
> > +        for dev in key_dict:
> > +            try:
> > +                if dev in HOMEMATIC_DEVICES:
> > +                    for hm_element in HOMEMATIC_DEVICES[dev]:
> > +                        hm_element.link_homematic()
> > +                else:
> > +                    devices_not_created.append(dev)
> > +            # pylint: disable=broad-except
> > +            except Exception as err:
>
> Pyhomematic currently doesn't raise it's own exceptions. I've created an
> issue for that in my repo.
> Is this mandatory to get the code merged, or can we take care of that in
> the next iteration of our code? We are working on more device-support and a
> few other features, so there definitely will be follow-up pull requests.
> For the time being I would move the broad try/except-block into the part
> of link_homematic() where exceptions would occur.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/2368/files/dfe1b8d9344fbed665896bf6c2b65de8fbb5f280#r68488956>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABYJ2nHsbbWyz_6JoJKp8PYDDxOqRIbyks5qPTXSgaJpZM4I9jcJ>
> .
>